### DISCUSSION TITLE
Possible to use CSS-in-JS with Static Site Generation ?

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#36412] [Help]
Hello everyone i am very new to Next.js and trying to figure out an efficient way to style.

Using Styled-Component caused me to have Cumulative Layout Shifts, and I think it has to do with JS styles not being rendered server side ?

My page with Style-Component causes me to drop to 85/100 LightHouse Performance Score due to CLS.
The same page using Next.js css modules styles gets me to 100/100 with no CLS.

My issue being, i kind of really like CSS-in-JS and have used Emotion for my previous project.
Is there a way to get good performance score with SSG + CSS-in-JS ? If so, how to render the styles server side ?

Thank you !

**Comment by vadymshymko** [comment]
Hi. 
> Is there a way to get good performance score with SSG + CSS-in-JS ?

In short - yes!) But, it seems to me, this is too general a question, since there can be many reasons for the large value of CLS)

> If so, how to render the styles server side ?

Please, see [styled components docs about integration with nextjs](https://styled-components.com/docs/advanced#nextjs). Hope this help!)

**Comment by icyJoseph** [comment]
Hi,

Styled Components gains an additional benefit when used in SSR, which is a hybrid mode, where the styles are calculated on the server, and therefore exist in the initial HTML send to the client, but can also be dynamic once in the client.

You must implement something called `ServerStyleSheet` in the `_document` file, see [here](https://github.com/vercel/next.js/blob/canary/examples/with-styled-components-babel/pages/_document.js).

I assume the CLS comes from a site that doesn't do the above, but generates the styles on the browser, which means there's a flash of un-styled content, and then suddenly everything gets shuffled around with the client side calculated styles.

**Reply by trompx** [reply]
So as you add getInitialProps to _document, you opt out of static optimization, so all your pages will be server side rendered, no more SSG?

**Reply by icyJoseph** [reply]
In _document it doesn't do that, only in _app, and under certain conditions.

**Reply by ashirbad29** [reply]
@icyJoseph can you please reshare the link for the example, I'm getting 404 on that now!

**Reply by icyJoseph** [reply]
https://github.com/vercel/next.js/blob/canary/examples/with-styled-components/pages/_document.tsx was it that one?
